core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java

```diff
@@ -0,0 +1,690 @@
+
+  /**
+   * Write segment information to the segment folder with indexfilename and
+   * corresponding partitions.
+   */
+  public void writeSegmentFile(String tablePath, final String taskNo, String location,
+      String timeStamp, List<String> partionNames) throws IOException {
+    String tempFolderLoc = timeStamp + ".tmp";
+    String writePath = CarbonTablePath.getSegmentFilesLocation(tablePath) + "/" + tempFolderLoc;
+    CarbonFile carbonFile = FileFactory.getCarbonFile(writePath);
+    if (!carbonFile.exists()) {
+      carbonFile.mkdirs(writePath, FileFactory.getFileType(writePath));
+    }
+    CarbonFile tempFolder =
+        FileFactory.getCarbonFile(location + CarbonCommonConstants.FILE_SEPARATOR + tempFolderLoc);
+    boolean isRelative = false;
+    if (location.startsWith(tablePath)) {
+      location = location.substring(tablePath.length(), location.length());
+      isRelative = true;
+    }
+    if (tempFolder.exists() && partionNames.size() > 0) {
+      CarbonFile[] carbonFiles = tempFolder.listFiles(new CarbonFileFilter() {
+        @Override public boolean accept(CarbonFile file) {
+          return file.getName().startsWith(taskNo) && file.getName()
+              .endsWith(CarbonTablePath.INDEX_FILE_EXT);
+        }
+      });
+      if (carbonFiles != null && carbonFiles.length > 0) {
+        SegmentFile segmentFile = new SegmentFile();
+        Map<String, FolderDetails> locationMap = new HashMap<>();
+        FolderDetails folderDetails = new FolderDetails();
+        folderDetails.setRelative(isRelative);
+        folderDetails.setPartitions(partionNames);
+        folderDetails.setStatus(SegmentStatus.SUCCESS.getMessage());
+        for (CarbonFile file : carbonFiles) {
+          folderDetails.getFiles().add(file.getName());
+        }
+        locationMap.put(location, folderDetails);
+        segmentFile.setLocationMap(locationMap);
+        String path = writePath + "/" + taskNo + CarbonTablePath.SEGMENT_EXT;
+        // write segment info to new file.
+        writeSegmentFile(segmentFile, path);
+      }
+    }
+  }
+
+  /**
+   * Writes the segment file in json format
+   * @param segmentFile
+   * @param path
+   * @throws IOException
+   */
+  public void writeSegmentFile(SegmentFile segmentFile, String path) throws IOException {
+    AtomicFileOperations fileWrite =
+        new AtomicFileOperationsImpl(path, FileFactory.getFileType(path));
+    BufferedWriter brWriter = null;
+    DataOutputStream dataOutputStream = null;
+    Gson gsonObjectToWrite = new Gson();
+    try {
+      dataOutputStream = fileWrite.openForWrite(FileWriteOperation.OVERWRITE);
+      brWriter = new BufferedWriter(new OutputStreamWriter(dataOutputStream,
+          Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET)));
+
+      String metadataInstance = gsonObjectToWrite.toJson(segmentFile);
+      brWriter.write(metadataInstance);
+    } finally {
+      if (null != brWriter) {
+        brWriter.flush();
```

Comment:
flush can throw exception which will not close the fileWriter. Move brWriter.flush() to after write call.

Reply:
ok